### PR TITLE
Offset implementation

### PR DESCRIPTION
**Pull request by framos-gemini**
This version implements the functionality to send offsets to the TCS.

### CODE REVIEW

File: conf/common.mk [hunk 1]
```diff
@@ -87,6 +87,6 @@ TMP_DIST_DIR := /tmp/$(DIST_PACKAGE_NAME)
 %.o: %.cpp
 	@echo 'Building file: $<'
 	@echo 'Invoking $(OS) C++ Compiler'
-	$(CXX) $(INC_DIRS) -std=c++11 -g -O0 -Wall -fPIC -c -Wno-deprecated -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@:%.o=%.d)" -o"$@" "$<"
+	$(CXX) $(INC_DIRS) -pthread -std=c++11 -g -O0 -Wall -fPIC -c -Wno-deprecated -fmessage-length=0 -MMD -MP -MF"$(@:%.o=%.d)" -MT"$(@:%.o=%.d)" -o"$@" "$<"
```
Comment (cquiroz) [inline conf/common.mk]: Why do we need this?

File: giapi/GeminiUtil.h [hunk 2]
```diff
@@ -87,6 +87,48 @@ class GeminiUtil {
...
 	static int getTcsContext(TcsContext& ctx, long timeout) throw (GiapiException);
 
+       /** Function that allows an offset to be applied to the TCS. There are two types 
+	 * of the offsets that instruments should indicate. For example, offsets applied
+	 * during the acquisition and offsets applied during the Slow Guiding Correction. 
+	 * 
+	 * @param p value of the P offset. The unit must to be in arcsecs. 
+         * @param q value of the Q offset. The unit must to be in arcsecs. 
+         * @param offsetType Type of the offset to be applied. 
+         *                      0 -> Adquistion.
```
Comment (cquiroz) [inline giapi/GeminiUtil.h]: I guess we are not really passing 0/1 but an actual type

File: giapi/giapi.h [hunk 3]
```diff
@@ -212,6 +212,16 @@ namespace giapi {
 		} po;
 		double ao2t[6]; //Optical distortion coefficients (Not used to date)
 	};
+
+        /**
+         * Definition of offset types. Two types of offset are implemented, 
+         * offset during the adquisition and offset during Slow Guiding Correction.  
+        */
+	enum OffsetType {
```
Comment (cquiroz) [inline giapi/giapi.h:220]: Do you think we should follow this convention?
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps

File: src/examples/gmp.properties [hunk 4]
```diff
@@ -1,2 +1,3 @@
 #gmp.hostname=192.158.189.129
 gmp.hostname=127.0.0.1
+gmp.instrument=igrins2
```
Comment (cquiroz) [inline src/examples/gmp.properties:3]: What is this used for?

File: src/examples/epicsget
Comment (cquiroz) [inline src/examples/epicsget:1]: Do we want to include binaries?

File: src/examples/tcsApplyOffset.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+		if (status == status::ERROR) {
+			cout << "Can't apply the offset to TCS..." << endl;
+			return 0;
+		}
+
+		cout<<"Was the TCS OFFSET applied: " << (status == 1) << endl;
+
+	} catch (GiapiException &e) {
+		cout << "exceptionnnn " << endl;
```
Comment (cquiroz) [inline src/examples/tcsApplyOffset.cpp]: typo

File: src/examples/tcsApplyOffsetCallback.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+#include<unistd.h>
+
+using namespace giapi;
+
+using namespace std;
+
+bool wasOffsetApplied=false;
+
+void receiveMsg(int wasOffsetAplied, string msg) {
+	string res = (wasOffsetAplied) ? "was applied " : " was NOT APPLIED. ";
```
Comment (cquiroz) [inline src/examples/tcsApplyOffsetCallback.cpp]: should this be uniform on its use of capitals?

File: src/gemini/tcs/jms/JmsApplyOffset.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+#include <src/util/StringUtil.h>
+#include <stdlib.h>
+#include <thread>
+#include<unistd.h>
+
+using namespace gmp;
+
+namespace giapi {
+
+namespace gemini {
```
Comment (cquiroz) [inline src/gemini/tcs/jms/JmsApplyOffset.cpp]: I don't know enough of C++11. Is not indenting namespaces a convention?

File: src/gemini/tcs/jms/JmsApplyOffset.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace tcs {
+
+namespace jms {
+
+JmsApplyOffset::JmsApplyOffset() throw (CommunicationException) :
+                                 JmsProducer(GMPKeys::GMP_TCS_OFFSET_DESTINATION) {
+	instName = giapi::util::PropertiesUtil::Instance().getProperty("gmp.instrument");
```
Comment (cquiroz) [inline src/gemini/tcs/jms/JmsApplyOffset.cpp]: We may want to document this on Giapi Design and Use

File: src/gemini/tcs/jms/JmsApplyOffset.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	int wasOffsetApplied = 0;
+	try {
+		//Create a message to do the request.
+		rMsg = _session->createBytesMessage();
+		//create temporary objects to get the answer
+		TemporaryQueue * tmpQueue = _session->createTemporaryQueue();
+		MessageConsumer * tmpConsumer = _session->createConsumer(tmpQueue);
+
+		//define the destination for the service to provide an answer
+		rMsg->setCMSReplyTo(tmpQueue);
```
Comment (cquiroz) [inline src/gemini/tcs/jms/JmsApplyOffset.cpp]: Cool, this is the protocol to send the request. I guess there is a counterpart on the GMP that will decode this and apply the offset on the TCS.
Is that correct?

**Review by cquiroz — COMMENTED** [review]
Thanks, it looks good, I added a few comments but it seems fine to me

File: giapi/giapi.h (hunk 3 above)
Comment (framos-dev) [reply giapi/giapi.h:220]: I 100% agree with your point of view. The only problem that I have seen all the [remains](https://github.com/gemini-hlsw/giapi-glue-cc/blob/53e5f84180c9aa5ae7e4cf9fe9f0076b800d5f03/giapi/giapi.h#L94C8-L94C23) enum declarations are using ALL_CAPS style. I think if we change it we are going to be mixing two different styles. I think the best solution would be to make a specific task to change to this style. What do you think about it?

Comment (cquiroz) [reply giapi/giapi.h:220]: Oh we have other enums with caps?
Then it is better to be consistent

File: src/examples/tcsApplyOffsetCallback.cpp [hunk 10]
```diff
@@ -13,11 +13,10 @@ using namespace std;
 
 bool wasOffsetApplied=false;
 
-void receiveMsg(int wasOffsetAplied, string msg) {
-	string res = (wasOffsetAplied) ? "was applied " : " was NOT APPLIED. ";
+void receiveMsg(int wasAplied, string msg) {
```
Comment (cquiroz) [inline src/examples/tcsApplyOffsetCallback.cpp:16]: `wasApplied`?

**Review by cquiroz — APPROVED** [review]
Looks good to me, thanks

**Pull request merged**